File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.py
class Solution:
     def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            nums1,nums2 = nums2,nums1
        len_x = len(nums1)
        len_y = len(nums2)
        low_val = 0
        high_val = len_x

        while(low_val <= high_val):
            x_mid = int((low_val + high_val)/2)
            y_sep = int((len_x+len_y+1)/2 - x_mid)
            if x_mid == 0:
                x_left_max = -math.inf
            else:
                x_left_max = nums1[x_mid-1]

            if x_mid == len_x:
                x_right_min = math.inf
            else:
                x_right_min = nums1[x_mid]
            if y_sep == 0:
                y_max_left = -math.inf
            else:
                y_max_left = nums2[y_sep-1]
            if y_sep == len_y:
                y_min_right = math.inf
            else:
                y_min_right = nums2[y_sep]

            if x_left_max <= y_min_right and y_max_left <= x_right_min:
                if ((len_x+len_y)%2) == 0:
                    return ((max(x_left_max,y_max_left) + min(x_right_min,y_min_right))/2)
                else:
                    return(max(x_left_max,y_max_left))
            elif x_left_max > y_min_right:
                high_val = x_mid - 1
            else:
                low_val = x_mid + 1

File: 4-median-of-two-sorted-arrays/4-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        // write your code here
        int len = nums1.length + nums2.length;
        if (len % 2 == 1) {
            return finKth(nums1, 0, nums2, 0, len / 2 + 1);
        } else {
            return (
                finKth(nums1, 0, nums2, 0, len / 2) + finKth(nums1, 0, nums2, 0, len / 2 + 1)
                ) / 2.0;
        }
    }
    
    private double finKth(int[] nums1, int indA, int[] nums2, int indB, int k) {
        if (indA >= nums1.length) {
            return nums2[indB + k - 1];
        }
        if (indB >= nums2.length) {
            return nums1[indA + k - 1];
        }
        if (k == 1) {
            return Math.min(nums1[indA], nums2[indB]);
        }
        
        int A_key = indA + k / 2 - 1 < nums1.length 
                    ? nums1[indA + k / 2 - 1]
                    : Integer.MAX_VALUE;
        int B_key = indB + k / 2 - 1 < nums2.length
                    ? nums2[indB + k / 2 - 1]
                    : Integer.MAX_VALUE;
        if (A_key > B_key) {
            return finKth(nums1, indA, nums2, indB + k / 2, k - k / 2);
        } else {
            return finKth(nums1, indA + k / 2, nums2, indB, k - k / 2);
        }
    }
}